File: src/apps/business-os-mobile/shared/launch.mjs
import { ANDROID_ENTRY, ANDROID_ORIGIN } from "./constants.mjs";

function scriptSafeJson(value) {
  const lineSeparator = new RegExp(String.fromCharCode(0x2028), "g");
  const paragraphSeparator = new RegExp(String.fromCharCode(0x2029), "g");
  return JSON.stringify(value)
    .replace(/</g, "\\u003c")
    .replace(/>/g, "\\u003e")
    .replace(/&/g, "\\u0026")
    .replace(lineSeparator, "\\u2028")
    .replace(paragraphSeparator, "\\u2029");
}

export function buildLaunchContext(metadata, secrets, platform = "mobile") {
  if (!metadata || !secrets?.password || !secrets?.capabilityToken) throw new Error("launch secrets are unavailable");
  if (metadata.capabilityExpiresAtMs <= Date.now()) throw new Error("pairing authorization is expired");
  const session = {
    authenticated: true,
    source: `${platform}_invite`,
    capability_token: secrets.capabilityToken,
    capability_expires_at_ms: metadata.capabilityExpiresAtMs,
    user: {
      id: metadata.sessionUser.id,
      display_name: metadata.sessionUser.displayName,
      role: metadata.sessionUser.role,
      is_admin: ["chef", "admin", "founder"].includes(metadata.sessionUser.role),
    },
  };
  const config = {
    instance_id: metadata.instanceId,
    peer_id: `${platform}:${metadata.id}`,
    peer_role: "business_os_client",
    native_peer_id: metadata.nativePeerId,
    sync_room: metadata.syncRoom,
    signaling_urls: metadata.signalingUrls,
    signaling_room_password: secrets.password,
    transport: "webrtc",
    data_plane: "rxdb-webrtc",
    http_bridge_available: false,
    app_hosting: `${platform}_bundled_shell`,
    ctox_instance_required: true,
    session,
  };
  return { session, config, designTemplates: [] };
}

export function injectLaunchContext(html, context) {
  const marker = /<head(?:\s[^>]*)?>/i;
  const match = marker.exec(String(html));
  if (!match) throw new Error("shell index is missing head");
  const script = `<script data-ctox-mobile-bootstrap>window.CTOX_BUSINESS_OS_SESSION=${scriptSafeJson(context.session)};window.CTOX_BUSINESS_OS_CONFIG=${scriptSafeJson(context.config)};window.CTOX_BUSINESS_OS_DESIGN_TEMPLATES=[];</script>`;
  return `${html.slice(0, match.index + match[0].length)}${script}${html.slice(match.index + match[0].length)}`;
}

export function iosEntry(instanceId) {
  return `ctox-business-os-mobile://${encodeURIComponent(instanceId)}/business-os/index.html`;
}

export function navigationDecision(raw, { platform, instanceId } = {}) {
  let url;
  try { url = new URL(raw); } catch { return "deny"; }
  if (platform === "ios" && url.protocol === "ctox-business-os-mobile:" && url.hostname === instanceId && url.pathname.startsWith("/business-os/")) return "allow";
  if (platform === "android" && url.origin === ANDROID_ORIGIN && url.pathname.startsWith("/business-os/")) return "allow";
  if (url.protocol === "https:") return "external";
  return "deny";
}

export { ANDROID_ENTRY };

File: src/apps/business-os-mobile/ios/Sources/CTOXMobileCore/Launch.swift
import Foundation

public struct LaunchContext: Sendable {
    public let session: [String: Sendable]
    public let config: [String: Sendable]
}

public enum LaunchBuilder {
    public static func context(instance: MobileInstance, password: String, capabilityToken: String, now: Date = Date()) throws -> (session: [String: Any], config: [String: Any]) {
        guard !password.isEmpty, !capabilityToken.isEmpty, instance.capabilityExpiresAt > now else { throw MobileInviteError.capabilityExpired }
        let isAdmin = ["chef", "admin", "founder"].contains(instance.sessionUser.role)
        let session: [String: Any] = [
            "authenticated": true,
            "source": "ios_invite",
            "capability_token": capabilityToken,
            "capability_expires_at_ms": Int64(instance.capabilityExpiresAt.timeIntervalSince1970 * 1000),
            "user": ["id": instance.sessionUser.id, "display_name": instance.sessionUser.displayName, "role": instance.sessionUser.role, "is_admin": isAdmin],
        ]
        let config: [String: Any] = [
            "instance_id": instance.instanceID,
            "peer_id": "ios:\(instance.id)",
            "peer_role": "business_os_client",
            "native_peer_id": instance.nativePeerID,
            "sync_room": instance.syncRoom,
            "signaling_urls": instance.signalingURLs,
            "signaling_room_password": password,
            "transport": "webrtc",
            "data_plane": "rxdb-webrtc",
            "http_bridge_available": false,
            "app_hosting": "ios_bundled_shell",
            "ctox_instance_required": true,
            "session": session,
        ]
        return (session, config)
    }

    public static func inject(html: Data, session: [String: Any], config: [String: Any]) throws -> Data {
        guard var text = String(data: html, encoding: .utf8), let range = text.range(of: #"<head(?:\s[^>]*)?>"#, options: [.regularExpression, .caseInsensitive]) else { throw CocoaError(.fileReadCorruptFile) }
        func encoded(_ value: Any) throws -> String {
            let data = try JSONSerialization.data(withJSONObject: value)
            return (String(data: data, encoding: .utf8) ?? "null")
                .replacingOccurrences(of: "<", with: "\\u003c")
                .replacingOccurrences(of: ">", with: "\\u003e")
                .replacingOccurrences(of: "&", with: "\\u0026")
        }
        let clipboardHardening = "try{Object.defineProperty(navigator,\u{27}clipboard\u{27},{value:{read:()=>Promise.reject(new DOMException(\u{27}Denied\u{27},\u{27}NotAllowedError\u{27})),readText:()=>Promise.reject(new DOMException(\u{27}Denied\u{27},\u{27}NotAllowedError\u{27})),write:()=>Promise.reject(new DOMException(\u{27}Denied\u{27},\u{27}NotAllowedError\u{27})),writeText:()=>Promise.reject(new DOMException(\u{27}Denied\u{27},\u{27}NotAllowedError\u{27}))},configurable:false})}catch(_){}"
        let script = "<script data-ctox-mobile-bootstrap>window.CTOX_BUSINESS_OS_SESSION=\(try encoded(session));window.CTOX_BUSINESS_OS_CONFIG=\(try encoded(config));window.CTOX_BUSINESS_OS_DESIGN_TEMPLATES=[];\(clipboardHardening)</script>"
        text.insert(contentsOf: script, at: range.upperBound)
        return Data(text.utf8)
    }
}

public enum NavigationPolicy: Equatable, Sendable { case allow, external, deny }
public enum MobileNavigation {
    public static func decision(for url: URL, instanceID: String) -> NavigationPolicy {
        if url.scheme == "ctox-business-os-mobile", url.host == instanceID, url.path.hasPrefix("/business-os/") { return .allow }
        if url.scheme == "https" { return .external }
        return .deny
    }
}
